### PR TITLE
Encapsulated MSP

### PR DESCRIPTION
**Pull request by wvarty**
This PR includes @JyeSmith 's MSP-OTA branch changes, which replaces the old switch ELRS opcode with a new one for MSP.

It also includes the changes in the CRSF lib to send encapsulated MSP crsf packets to betaflight.

### CODE REVIEW

File: src/src/tx_main.cpp [hunk 1]
```diff
@@ -343,22 +350,16 @@ void ICACHE_RAM_ATTR SendRCdataToRF()
   }
   else
   {
-    #if defined HYBRID_SWITCHES_8
```
Comment (JyeSmith) [inline src/src/tx_main.cpp:346]: I think HYBRID_SWITCHES_8 and SEQ_SWITCHES need to be added back, but moved to the else statement with Generate4ChannelData_11bit();

File: src/src/targets.h [hunk 2]
```diff
@@ -12,7 +12,7 @@
 /// General Features ///
 #define FEATURE_OPENTX_SYNC //uncomment to use OpenTX packet sync feature (requires OpenTX 2.4 onwards) - this reduces latency.
 #define LED_MAX_BRIGHTNESS 50 //0..255 for max led brightness
-
+// #define USE_UART2 // uncomment to use UART2 for MSP on the TX module
```
Comment (JyeSmith) [inline src/src/targets.h:15]: Should we stick this in user_defines.txt?

File: src/src/tx_main.cpp (hunk 1 above)
Comment (wvarty) [reply src/src/tx_main.cpp:346]: Something screwy has happened here with the merge from Master. I'll fix it up.

File: src/src/targets.h (hunk 2 above)
Comment (wvarty) [reply src/src/targets.h:15]: Good idea!

**Comment by wvarty** [comment]
@JyeSmith thanks for the thorough review, good pickups / suggestions. I'll make these changes tonight.

File: src/lib/CRSF/CRSF.cpp [hunk 3]
```diff
@@ -383,6 +383,37 @@ void ICACHE_RAM_ATTR CRSF::sendSyncPacketToTX(void *pvParameters) // in values i
...
+            // Encapsulated MSP payload
+            outBuffer[5] = 0x30; // header
+            outBuffer[6] = packet->payloadSize; // mspPayloadSize
+            outBuffer[7] = packet->function; // packet->cmd
+            for (uint8_t i = 0; i < ENCAPSULATED_MSP_PAYLOAD_SIZE; ++i) {
+                // copy packet payload into outBuffer and pad with zeros where required
+                outBuffer[8 + i] = i < packet->payloadSize ? packet->payload[i] : 0;
+            }
+            // Encapsulated MSP crc
+            outBuffer[12] =  CalcCRCMsp(&outBuffer[6], ENCAPSULATED_MSP_FRAME_LEN - 2);
```
Comment (JBKingdon) [inline src/lib/CRSF/CRSF.cpp:410]: The 12 and 13 constants are a bit of a hostage to fortune. Mighe be worth calculating a constant for the total length around line 391 before allocating the array, and then use total length -2 and total length -1 as the indices here?

File: src/src/tx_main.cpp [hunk 4]
```diff
@@ -209,20 +211,25 @@ void ICACHE_RAM_ATTR Generate4ChannelData_11bit()
...
+  Radio.TXdataBuffer[5] = 0;
+  Radio.TXdataBuffer[6] = 0;
+  if (MSPPacket.payloadSize <= 4)
+  {
+    MSPPacket.payloadReadIterator = 0;
+    for (int i = 0; i < MSPPacket.payloadSize; i++)
+    {
+      Radio.TXdataBuffer[3+i] = MSPPacket.readByte();
+    }
+  }
```
Comment (JBKingdon) [inline src/src/tx_main.cpp:232]: do we want a debug output on the 'else' side of this if statement?

File: src/src/tx_main.cpp [hunk 5]
```diff
@@ -209,20 +211,25 @@ void ICACHE_RAM_ATTR Generate4ChannelData_11bit()
...
-  Radio.TXdataBuffer[4] = Radio.TXdataBuffer[2];
-  Radio.TXdataBuffer[5] = Radio.NonceTX;
-  Radio.TXdataBuffer[6] = FHSSgetCurrIndex();
+  Radio.TXdataBuffer[1] = MSPPacket.function;
+  Radio.TXdataBuffer[2] = MSPPacket.payloadSize;
+  Radio.TXdataBuffer[3] = 0;
+  Radio.TXdataBuffer[4] = 0;
+  Radio.TXdataBuffer[5] = 0;
+  Radio.TXdataBuffer[6] = 0;
+  if (MSPPacket.payloadSize <= 4)
```
Comment (JBKingdon) [inline src/src/tx_main.cpp:225]: do we want a debug output on the 'else' side of this if statement?

File: src/user_defines.txt [hunk 6]
```diff
@@ -43,6 +43,8 @@
 
 #-DHYBRID_SWITCHES_8
 
+# Uncomment to use UART2 for MSP comms on the TX module
```
Comment (JBKingdon) [inline src/user_defines.txt:46]: Thanks for this! :)

File: src/src/tx_main.cpp [hunk 7]
```diff
@@ -350,16 +350,13 @@ void ICACHE_RAM_ATTR SendRCdataToRF()
   }
   else
   {
-    if ((millis() > (MSP_PACKET_SEND_INTERVAL + MSPPacketLastSent)) && MSPPacketSendCount)
```
Comment (JyeSmith) [inline src/src/tx_main.cpp]: We need this.

File: src/lib/CRSF/CRSF.cpp (hunk 3 above)
Comment (wvarty) [reply src/lib/CRSF/CRSF.cpp]: Changed based on suggestions.

File: src/src/tx_main.cpp (hunk 5 above)
Comment (wvarty) [reply src/src/tx_main.cpp:225]: Added

File: src/src/tx_main.cpp (hunk 7 above)
Comment (wvarty) [reply src/src/tx_main.cpp]: Added. Please double triple check this bit ;)

**Comment by JBKingdon** [comment]
Looks good to me :)

**Pull request merged**